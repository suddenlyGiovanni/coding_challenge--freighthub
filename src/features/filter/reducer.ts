import { Reducer } from 'redux'
import { getType, ActionType } from 'typesafe-actions'

import * as filterActions from './actions'
import * as FilterTypes from './typings'

export type FilterAction = ActionType<typeof filterActions>

export type FilterState = Readonly<FilterTypes.Filter>

const initialState: FilterState = {
  id: '',
  mode: {
    air: true,
    sea: true,
    rail: true,
  },
  status: 'ALL',
}

export const filterReducer: Reducer<FilterState, FilterAction> = (
  state = initialState,
  action
) => {
  if (action.type === getType(filterActions.setFilter)) {
    const { id, mode, status } = action.payload
    return {
      ...state,
      id: id === undefined ? state.id : id,
      status: status || state.status,
      mode: mode
        ? {
            ...state.mode,
            ...mode,
          }
        : state.mode,
    }
  }
  return state
}
