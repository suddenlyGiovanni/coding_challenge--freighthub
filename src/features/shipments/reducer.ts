import { Reducer } from 'redux'
import { getType, ActionType } from 'typesafe-actions'

import * as shipmentsActions from './actions'
import * as ShipmentsTypes from './typings'

export type ShipmentsAction = ActionType<typeof shipmentsActions>

export type ShipmentsState = Readonly<ShipmentsTypes.Shipments>

const initialState: ShipmentsState = []

export const shipmentsReducer: Reducer<ShipmentsState, ShipmentsAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case getType(shipmentsActions.setShipments): {
      return [...action.payload.shipments]
    }

    case getType(shipmentsActions.editShipmentName): {
      const { id, name } = action.payload
      return state.map(storeShipment => {
        return storeShipment.id === id
          ? { ...storeShipment, name } //
          : storeShipment //
      })
    }

    case getType(shipmentsActions.setShipment): {
      const { shipment } = action.payload
      return state.map(storeShipment => {
        return storeShipment.id === shipment.id
          ? shipment //
          : storeShipment //
      })
    }

    default:
      return state
  }
}
