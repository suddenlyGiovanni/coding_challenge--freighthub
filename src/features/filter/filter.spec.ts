import { RootState } from 'typesafe-actions'
import {
  filterActions as actions,
  filterTypes as types,
  filterReducer as reducer,
  filterSelectors as selectors,
} from './index'

import { FilterState } from './reducer'

/**
 * FIXTURES
 */
function getInitialState(initial?: FilterState) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return reducer(initial, {} as any)
}

function getRootState(initial?: FilterState) {
  return { filter: getInitialState(initial) }
}

const filter = {
  id: 'U1000',
  mode: { air: false },
  status: 'ACTIVE' as const,
}

describe('feature: `filter`', () => {
  // ACTIONS CREATOR
  describe('actions', () => {
    it('setFilter should create an action with the correct signature', () => {
      const expectedActionShape = {
        type: types.SET_FILTER_SHIPMENTS,
        payload: {
          id: 'U1000',
          mode: { air: false },
          status: 'ACTIVE',
        },
      }
      expect(actions.setFilter(filter)).toMatchObject(expectedActionShape)
    })
  })

  // REDUCER
  describe('reducer', () => {
    it('initial state should match a snapshot', () => {
      const initialState = getInitialState()
      expect(initialState).toMatchSnapshot()
    })

    it('set the filter state to the store', () => {
      // Arrange
      const initialState = getInitialState()
      // Act
      const stateSnapshot1 = reducer(initialState, actions.setFilter(filter))
      const stateSnapshot2 = reducer(
        stateSnapshot1,
        actions.setFilter({ mode: { sea: false } })
      )

      const stateSnapshot3 = reducer(
        stateSnapshot2,
        actions.setFilter({ status: 'COMPLETED' })
      )
      // Assert
      expect(stateSnapshot1).toMatchObject<FilterState>({
        id: 'U1000',
        mode: {
          air: false,
          sea: true,
          rail: true,
        },
        status: 'ACTIVE',
      })

      expect(stateSnapshot2).toMatchObject<FilterState>({
        id: 'U1000',
        mode: {
          air: false,
          sea: false,
          rail: true,
        },
        status: 'ACTIVE',
      })

      expect(stateSnapshot3).toMatchObject<FilterState>({
        id: 'U1000',
        mode: {
          air: false,
          sea: false,
          rail: true,
        },
        status: 'COMPLETED',
      })
    })
  })

  // SELECTORS
  describe('selectors', () => {
    it('retrieves the filter branch form the store', () => {
      const rootState = getRootState()
      expect(selectors.getFilter(rootState as RootState)).toMatchObject({
        id: '',
        mode: {
          air: true,
          sea: true,
          rail: true,
        },
        status: 'ALL',
      })
    })
  })
})
