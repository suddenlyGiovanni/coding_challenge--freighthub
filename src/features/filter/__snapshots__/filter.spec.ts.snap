// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature: \`filter\` reducer initial state should match a snapshot 1`] = `
Object {
  "id": "",
  "mode": Object {
    "air": true,
    "rail": true,
    "sea": true,
  },
  "status": "ALL",
}
`;
